Write a blog post on Docker for Mac and Windows

tkw1536 opened the pull request:
In the scope of OpenDreamKit/OpenDreamKit#43 I have written a short blog post about Docker now being available for Mac and Windows. Before publishing it on the website I would like to have some feedback on it, also to see if that is what @defeo imagined for the deliverable or if I should go into a different direction.

tkw1536 committed:
Write a blog post on Docker for Mac and Windows

In the scope of OpenDreamKit/OpenDreamKit#43 I have written a short blog
post about Docker now being available natively for Mac and Windows.

olexandr-konovalov commented on _posts/2016-08-18-emerging-docker.md:
```diff
@@ -0,0 +1,16 @@
+
+Within the ODK project this allows us to bundle mathematical software systems (such as [GAP](http://www.gap-system.org/), [Sage](http://www.sagemath.org/), ...) and distribute them in a manner that is both accessible to users and easily maintainable for developers. Furthermore the isolated aspect of docker containers allows us to easily integrate multiple systems together without having to make a ...
+
+Docker was originally a linux only application -- it relied on a lot of functionality provided by the Linux kernel. To make it available on Windows and Mac the developers provided a virtual machine that ran a streamlined Linux system with Docker pre-installed. This made it possible for Windows and Mac users to run docker containers,  however it introduced an additional layer of abstraction that c ...
```
Here and below - I think `Linux` usually starts from the capital `L`

olexandr-konovalov commented on _posts/2016-08-18-emerging-docker.md:
```diff
@@ -0,0 +1,16 @@
+
+Docker was originally a linux only application -- it relied on a lot of functionality provided by the Linux kernel. To make it available on Windows and Mac the developers provided a virtual machine that ran a streamlined Linux system with Docker pre-installed. This made it possible for Windows and Mac users to run docker containers,  however it introduced an additional layer of abstraction that c ...
+
+Recently Docker started to build native versions for Windows and Mac. These versions do not rely on linux functionality -- instead they leverage functionality provided by Windows and Mac operating systems natively. In terms of the OpenDreamKit project these are a big step in terms of usability -- they make it significantly easier for users to run a Docker container. Users can now install Docker j ...
```
I would say "...  for Windows and OS X" since `OS X` is the official name of the operating system

olexandr-konovalov commented on _posts/2016-08-18-emerging-docker.md:
```diff
@@ -0,0 +1,16 @@
+
+Docker provides so-called software images that package software. These images can then be executed inside a docker container. Furthermore, Docker provides a service called [DockerHub](https://hub.docker.com/) that allows users to upload and publish their own images. This allows developers to bundle their software and distribute it easily. Furthermore, since each container runs isolated from the r ...
+
+Within the ODK project this allows us to bundle mathematical software systems (such as [GAP](http://www.gap-system.org/), [Sage](http://www.sagemath.org/), ...) and distribute them in a manner that is both accessible to users and easily maintainable for developers. Furthermore the isolated aspect of docker containers allows us to easily integrate multiple systems together without having to make a ...
```
Maybe replace `, ...` by "and others" or "etc." - otherwise looks too much like a draft

olexandr-konovalov commented:
Thanks @tkw1536! I've left 3 comments, otherwise looks good to me!

tkw1536 committed:
Fix typos in Docker for Windows and Mac blogpost

thanks to @alex-konovalov for finding them

tkw1536 commented:
Thanks, I have updated the pull request accordingly!

defeo commented:
Yup, looks good to me. I spotted a couple of minor language issues here and there. I'm not very experienced with GH pull requests, is there a way I can add a commit on top of this PR without merging into master?

tkw1536 commented:
You could clone locally and then commit it to master afterwards. Or you can fork my fork. Or you can commit in the actual diff as alex did above.

Commit:
Minor corrections

defeo commented:
Well committing to master is what I wanted to avoid. Whatever, here's the diff, I suppose you know how to use `git am`. Do whatever you feel like with it.

``` diff
From 3e9fb453a971d74f5bd7e0a2ee472a36890016ab Mon Sep 17 00:00:00 2001
From: Luca De Feo <defeo@pippin>
Date: Thu, 18 Aug 2016 15:49:31 +0200
Subject: [PATCH] Minor corrections

---
 _posts/2016-08-18-emerging-docker.md | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/_posts/2016-08-18-emerging-docker.md b/_posts/2016-08-18-emerging-docker.md
index b8b86e1..de8f877 100644
--- a/_posts/2016-08-18-emerging-docker.md
+++ b/_posts/2016-08-18-emerging-docker.md
@@ -7,10 +7,10 @@ title: Running Docker Natively on Windows and Mac

 Docker provides so-called software images that package software. These images can then be executed inside a docker container. Furthermore, Docker provides a service called [DockerHub](https://hub.docker.com/) that allows users to upload and publish their own images. This allows developers to bundle their software and distribute it easily. Furthermore, since each container runs isolated from the rest of the systems, the developers do not have to rely on any kind of other system configuration.

-Within the ODK project this allows us to bundle mathematical software systems (such as [GAP](http://www.gap-system.org/), [Sage](http://www.sagemath.org/), etc) and distribute them in a manner that is both accessible to users and easily maintainable for developers. Furthermore the isolated aspect of docker containers allows us to easily integrate multiple systems together without having to make additional assumptions about the users specific setup -- we can just run all systems in one docker container. This way users can install the entire mathematical software stack that the OpenDreamKit project aims to provide easily.
+Within the OpenDreamKit project this allows us to bundle mathematical software systems (such as [GAP](http://www.gap-system.org/), [Sage](http://www.sagemath.org/), etc) and distribute them in a manner that is both accessible to users and easily maintainable for developers. Furthermore the isolated aspect of docker containers allows us to easily integrate multiple systems together without having to make additional assumptions about the user's specific setup -- we can just run all systems in one docker container. This way users can install the entire mathematical software stack that the OpenDreamKit project aims to provide easily.

-Docker was originally a Linux only application -- it relied on a lot of functionality provided by the Linux kernel. To make it available on Windows and Mac the developers provided a virtual machine that ran a streamlined Linux system with Docker pre-installed. This made it possible for Windows and Mac users to run docker containers,  however it introduced an additional layer of abstraction that came with some disadvantages. The additional virtualisation slowed down docker containers and provided technical limitations when wanting to integrate with the host system. It also required users to install virtualisation software before being able to run any kind of Docker Image -- something that was not very user friendly.
+Docker was originally a Linux only application -- it relied on a lot of functionality provided by the Linux kernel. To make it available on Windows and Mac the developers provide a virtual machine, called Docker Toolbox, that runs a streamlined Linux system with Docker pre-installed. This makes it possible for Windows and Mac users to run docker containers,  however it introduces an additional layer of abstraction that comes with some disadvantages. The additional virtualisation slows down docker containers and faces technical limitations when wanting to integrate with the host system. It also requires users to install virtualisation software before being able to run any kind of Docker Image -- something that is not very user friendly.

-Recently Docker started to build native versions for Windows and Mac These versions do not rely on Linux functionality -- instead they leverage functionality provided by Windows and OS X operating systems natively. In terms of the OpenDreamKit project these are a big step in terms of usability -- they make it significantly easier for users to run a Docker container. Users can now install Docker just as they would install any other software on their machine. Among speed and resource advantages, these versions should make it easier for developers to create docker-powered applications and Docker containers because of better integration between host system and containers,  for example the file system of the physical machine can be mounted inside containers more easily. As a side note, the windows version of Docker still requires some manual configuration -- a setting in the BIOS has to be changed. This is much less effort than was required previously, however it would be much easier if users did not have to do any manual configuration.
+Recently Docker started to build native versions for Windows and Mac. These versions do not rely on Linux functionality -- instead they leverage functionality provided by Windows and OS X operating systems natively. In terms of the OpenDreamKit project these are a big step in terms of usability -- they make it significantly easier for users to run a Docker container. Users can now install Docker just as they would install any other software on their machine. Among speed and resource advantages, these versions will make it easier for developers to create docker-powered applications and Docker containers because of better integration between host system and containers,  for example the file system of the physical machine can be mounted inside containers more easily. As a side note, the Windows version of Docker still requires some manual configuration -- a setting in the BIOS has to be changed. This is much less effort than was required previously, however it would be much easier if users did not have to do any manual configuration.

 Docker for Windows and Mac has been in a private Beta since March 2016 and has recently become available as a public Beta. Interested readers can find more information on the [Docker Blog](https://blog.docker.com/2016/06/docker-mac-windows-public-beta/). A more detailed introduction to Docker can also be found in the [Docker Documentation](https://docs.docker.com/engine/understanding-docker/).
-- 
2.9.2
```

defeo commented:
Two more things.

> It also required users to install virtualisation software before being able to run any kind of Docker Imag

Are you sure about this? I thought VirtualBox got installed along with the image.

> As a side note, the windows version of Docker still requires some manual configuration -- a setting in the BIOS has to be changed.

Could you say a little more about this? A pointer to some reference would be enough.

kohlhase commented:
> As a side note, the windows version of Docker still requires some manual configuration -- a setting in the BIOS has to be changed.
> 
> Could you say a little more about this? A pointer to some reference would be enough.

I support this, we should have a place that tells users exactly what they need to do in windows, which windows versions work, ...

tkw1536 committed:
Clarifications to Blog Post

Clarify that VirtualBox is installed on Docker Machine for Windows / Mac
and clatify Docker for Windows caveats.

tkw1536 commented:
@defeo 

> Two more things.
> 
> > It also required users to install virtualisation software before being able to run any kind of Docker Imag
> 
> Are you sure about this? I thought VirtualBox got installed along with the image.

You are correct, I have clarified what I meant by that. 

@kohlhase 

> > > As a side note, the windows version of Docker still requires some manual configuration -- a setting in the BIOS has to be changed.
> > 
> > Could you say a little more about this? A pointer to some reference would be enough.
> 
> I support this, we should have a place that tells users exactly what they need to do in windows, which windows versions work, ...

Support is only for Windows 10 Pro & Enterprise, additionally Hyper-V has to be enabled. I have added a pointer to the documentation at https://docs.docker.com/docker-for-windows/#/what-to-know-before-you-install .

defeo commented on _posts/2016-08-18-emerging-docker.md:
```diff
@@ -0,0 +1,16 @@
+
+Within the OpenDreamKit project this allows us to bundle mathematical software systems (such as [GAP](http://www.gap-system.org/), [Sage](http://www.sagemath.org/), etc) and distribute them in a manner that is both accessible to users and easily maintainable for developers. Furthermore the isolated aspect of docker containers allows us to easily integrate multiple systems together without having  ...
+
+Docker was originally a Linux only application -- it relied on a lot of functionality provided by the Linux kernel. To make it available on Windows and Mac the developers provide a virtual machine, called [Docker Toolbox](https://www.docker.com/products/docker-toolbox), that runs a streamlined Linux system with Docker pre-installed. This makes it possible for Windows and Mac users to run docker c ...
```
I had moved all this paragraph to the present tense because Docker Toolbox is still distributed. However, the last sentence you just added is in past tense.

tkw1536 replied:
@defeo Sorry, that has been fixed.

tkw1536 committed:
Fix tense of blogpost

defeo commented:
It's unfortunate that there are no step-by-step instructions in the docker docs on how to enable Hyper-V in the BIOS. But, then again, docker for windows is supported only in Pro & Enterprise, so presumably you are a professional knowing what you're doing if you're installing it.

Hopefully things will get better in the future. That's what docker devs promise, at least. This message is important to our community, it is important that we deliver it clearly.

Concerning OS X, aren't there any gotchas?

tkw1536 commented:
None that I know of. For me and @kohlhase it installed without any hassle.

kohlhase commented:
> None that I know of. For me and @kohlhase it installed without any hassle.

I can confirm this. I am using the public beta on MacOS quite happily (getting lots of updates though, but it is a beta) since the Bremen meeting.

defeo commented:
That is really great news.

tkw1536 commented:
Can we merge this pull request then and consider this part of the report completed?

defeo commented:
If you don't mind, I'd like to keep this on hold until the end of the month, so we can coordinate with the other parts of D2.3. I do not exclude making some more minor edits.

tkw1536 commented:
Maybe make a d2-3 branch on this repository then? That way we could
coordinate easier.

On Thu, Aug 18, 2016, 18:26 Luca De Feo notifications@github.com wrote:

> If you don't mind, I'd like to keep this on hold until the end of the
> month, so we can coordinate with the other parts of D2.3. I do not exclude
> making some more minor edits.
> 
> —
> You are receiving this because you were mentioned.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/OpenDreamKit/OpenDreamKit.github.io/pull/33#issuecomment-240778261,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACya6kWGR_a516wP6_FOLxfiYBvPNElEks5qhIebgaJpZM4Jnb1u
> .

defeo commented:
This is now branch https://github.com/OpenDreamKit/OpenDreamKit.github.io/tree/d2-3-docker.

I'm integrating now in OpenDreamKit/OpenDreamKit#43.

If there are no obstacles, I will publish this on Fri on the website.

defeo commented:
I just published this. Thanks, @tkw1536.
